maketeams3.py

```diff
@@ -76,177 +69,197 @@ def getPlayer(self, board):
-        swapPlayers(*(swaps[-1][1]))
-        print(swaps[-1])
-        updatePref()
-        updateSort()
-        p = 0
-    else: # go to the next player in the list
-        p += 1
-
-for player in players:
-    player.setReqMet()
-
-#WIP for upload format for heltour
-jsonoutput = []
-#[{"action":"change-member","team_number":1,"board_number":1,"player":{"name":"lemonworld","is_captain":false,"is_vice_captain":false}}]
-for t, team in enumerate(teams):
-    for b, board in enumerate(team.boards):
-        pp = {"action":"change-member","team_number":t+1,"board_number":b+1,"player":{"name":board.name,"is_captain":False,"is_vice_captain":False}}
-        jsonoutput.append(pp)
-for b, board in enumerate(alts_split):
-    print(board)
-    for _, pp in enumerate(board):
-        pp = {"action":"create-alternate","board_number":b+1,"player_name":pp.name}
-        jsonoutput.append(pp)
-
-if output == "readable":
-    for team in teams:
-        print(team)
-    print("ALTERNATES")
-    for board in alts_split:
+@click.command()
+@click.option('--output', default="readable", type=click.Choice(['json', 'readable']))
+@click.option('--players', help='the json file containing the players.', required=True)
+def make_teams(players, output):
+    # input file is JSON data with the following keys: rating, name, in_slack, account_status, date_created, prefers_alt, friends, avoid, has_20_games.
+    with open(players,'r') as infile:
+        playerdata = json.load(infile)
+    print("This data was read from file.")
+
+    # put player data into Player objects
+    players = []
+    for player in playerdata:
+        if player['has_20_games'] and player['in_slack']:
+            players.append(Player(player['name'], player['rating'], player['friends'], player['avoid'], player['date_created'], player['prefers_alt']))
+        else:
+            print("{0} skipped".format(player['name']))
+    players.sort(key=lambda player: player.rating, reverse=True)
+
+    # splits list of Player objects into 6 near equal lists, sectioned by rating
+    avg = len(players) / 6.0
+    print(len(players))
+    players_split = []
+    last = 0.0
+    while round(last) < len(players):
+        players_split.append(players[int(round(last)):int(round(last + avg))])
+        last += avg
+
+    num_teams = int(math.ceil((len(players_split[0])*0.8)/2.0)*2)
+    print(num_teams)
+
+    # separate preferred alternates into alternates lists
+    alts_split = [[] for i in range(6)]
+    for n, board in enumerate(players_split):
+        for player in board:
+            if player.alt:
+                alts_split[n].append(player)
+    for n, board in enumerate(alts_split):
+        for alt in board:
+            players_split[n].remove(alt)
+
+    # ensure the number of teams doesn't exceed the number of players available
+    # on a given board
+    num_teams = min([num_teams] + [len(board) for board in players_split])
```

Comment:
If I follow, this will reduce the number of teams to accommodate the number of people who want to be alternates. Unfortunately this forces more players to be alternates to fill the other spaces. I think better would be either to re-balance the players who request to be alternates between the boards so there are not too many on one board - or what we've done manually before is change a player who requested alternate to be full time (after asking them).